File: data/cpp/tests/test_Whoosh_searcher_search.cpp
#include <gtest/gtest.h>
#include <string>
#include <filesystem>
#include <fstream>
#include "../src/function_Whoosh_searcher_search.cpp"

namespace fs = std::filesystem;

class TestWhooshSearch : public ::testing::Test {
protected:
    std::string index_dir = "test_search_index";
    WhooshSearcher searcher;

    void SetUp() override {
        create_index(index_dir);
    }

    void TearDown() override {
        if (fs::exists(index_dir)) {
            fs::remove_all(index_dir);
        }
    }

    void create_index(const std::string& index_dir) {
        if (!fs::exists(index_dir)) {
            fs::create_directory(index_dir);
        }

        Xapian::WritableDatabase db(index_dir, Xapian::DB_CREATE_OR_OPEN);

        // 添加文档
        Xapian::Document doc1;
        doc1.set_data("This is a test document.");
        doc1.add_term("test");
        doc1.add_term("document");
        db.add_document(doc1);

        Xapian::Document doc2;
        doc2.set_data("Another test document.");
        doc2.add_term("test");
        doc2.add_term("document");
        db.add_document(doc2);

        Xapian::Document doc3;
        doc3.set_data("Document with a different content.");
        doc3.add_term("document");
        doc3.add_term("content");
        db.add_document(doc3);
    }
};

TEST_F(TestWhooshSearch, test_search_document) {
    std::string result = searcher.perform_search(index_dir, "test");
    EXPECT_EQ(result, "This is a test document.,Another test document.");
}

TEST_F(TestWhooshSearch, test_search_no_results) {
    std::string result = searcher.perform_search(index_dir, "nonexistent");
    EXPECT_EQ(result, "No documents found");
}

TEST_F(TestWhooshSearch, test_search_multiple_matches) {
    std::string result = searcher.perform_search(index_dir, "document");
    EXPECT_EQ(result, "This is a test document., Document with a different content.,Another test document.");
}

TEST_F(TestWhooshSearch, test_empty_query) {
    std::string result = searcher.perform_search(index_dir, "");
    EXPECT_EQ(result, "No documents found");
}

TEST_F(TestWhooshSearch, test_invalid_index_directory) {
    EXPECT_THROW(searcher.perform_search("invalid_dir", "test"), std::invalid_argument);
}

File: data/java/tests/test_Whoosh_searcher_search.java
package com.example;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Comparator;

import static org.junit.jupiter.api.Assertions.*;

class test_Whoosh_searcher_search {

    private String indexDir = "test_search_index";
    private function_Whoosh_searcher_search searcher = new function_Whoosh_searcher_search();

    @BeforeEach
    void setUp() throws IOException {
        createIndex(indexDir);
    }

    @AfterEach
    void tearDown() throws IOException {
        deleteDirectory(Paths.get(indexDir));
    }


    void createIndex(String indexDir) throws IOException {
        Path indexPath = Paths.get(indexDir);
        if (!Files.exists(indexPath)) {
            Files.createDirectories(indexPath);
        }

        org.apache.lucene.store.Directory directory = org.apache.lucene.store.FSDirectory.open(indexPath);
        org.apache.lucene.analysis.Analyzer analyzer = new org.apache.lucene.analysis.standard.StandardAnalyzer();
        org.apache.lucene.index.IndexWriterConfig config = new org.apache.lucene.index.IndexWriterConfig(analyzer);
        org.apache.lucene.index.IndexWriter writer = new org.apache.lucene.index.IndexWriter(directory, config);

        org.apache.lucene.document.Document doc1 = new org.apache.lucene.document.Document();
        doc1.add(new org.apache.lucene.document.TextField("content", "This is a test document.", org.apache.lucene.document.Field.Store.YES));
        writer.addDocument(doc1);

        org.apache.lucene.document.Document doc2 = new org.apache.lucene.document.Document();
        doc2.add(new org.apache.lucene.document.TextField("content", "Another test document.", org.apache.lucene.document.Field.Store.YES));
        writer.addDocument(doc2);

        org.apache.lucene.document.Document doc3 = new org.apache.lucene.document.Document();
        doc3.add(new org.apache.lucene.document.TextField("content", "Document with a different content.", org.apache.lucene.document.Field.Store.YES));
        writer.addDocument(doc3);

        writer.close();
        directory.close();
    }

    void deleteDirectory(Path path) throws IOException {
        if (Files.exists(path)) {
            Files.walk(path)
                    .sorted(Comparator.reverseOrder())
                    .map(Path::toFile)
                    .forEach(File::delete);
        }
    }

    @Test
    void testSearchDocument() {
        String results = searcher.perform_search(indexDir, "test");
        assertEquals("Another test document., This is a test document.", results);
    }

    @Test
    void testSearchNoResults() {
        String results = searcher.perform_search(indexDir, "nonexistent");
        assertEquals("No documents found", results);
    }

    @Test
    void testSearchMultipleMatches() {
        String results = searcher.perform_search(indexDir, "document");
        assertEquals("Another test document., Document with a different content., This is a test document.", results);
    }

    @Test
    void testEmptyQuery() {
        String results = searcher.perform_search(indexDir, "");
        assertEquals("No documents found", results);
    }

    @Test
    void testInvalidIndexDirectory() {
        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
            searcher.perform_search("invalid_dir", "test");
        });

        String expectedMessage = "Index directory does not exist: invalid_dir";
        String actualMessage = exception.getMessage();

        assertEquals(expectedMessage, actualMessage);
    }
}